src/main/java/com/puppycrawl/tools/checkstyle/utils/ChainedPropertyUtil.java

```diff
@@ -0,0 +1,134 @@
+     *                   for property resolution.
+     * @return resolved properties
+     * @throws CheckstyleException when chained property is not defined
+     */
+    public static Properties getResolvedProperties(Properties properties)
+            throws CheckstyleException {
+        final Properties resolvedProperties = new Properties();
+        final Set<String> unresolvedPropertyNames =
+            new HashSet<>(properties.stringPropertyNames());
+        Iterator<String> unresolvedPropertyIterator = unresolvedPropertyNames.iterator();
+        int numberOfPropertiesLeftToResolve = unresolvedPropertyNames.size();
+
+        while (unresolvedPropertyIterator.hasNext()) {
+            final String propertyName = unresolvedPropertyIterator.next();
+            String propertyValue = properties.getProperty(propertyName);
+            final Matcher matcher = PROPERTY_VARIABLE_PATTERN.matcher(propertyValue);
+
+            while (matcher.find()) {
+                final String propertyVariableExpression = matcher.group();
+                final String unresolvedPropertyName =
+                    getPropertyNameFromExpression(propertyVariableExpression);
+
+                final String resolvedPropertyValue =
+                    resolvedProperties.getProperty(unresolvedPropertyName);
+
+                if (resolvedPropertyValue != null) {
+                    propertyValue = propertyValue.replace(propertyVariableExpression,
+                        resolvedPropertyValue);
+                }
+            }
+
+            if (allChainedPropertiesAreResolved(propertyValue)) {
+                resolvedProperties.setProperty(propertyName, propertyValue);
+                unresolvedPropertyIterator.remove();
+            }
+
+            if (!unresolvedPropertyIterator.hasNext()) {
+                final boolean isUnresolvablePropertyPresent =
+                    numberOfPropertiesLeftToResolve == unresolvedPropertyNames.size();
+                if (isUnresolvablePropertyPresent) {
+                    // At this point, we will have not resolved any properties in two iterations,
+                    // so unresolvable properties exist.
+                    throw new CheckstyleException(UNDEFINED_PROPERTY_MESSAGE
+                        + unresolvedPropertyNames);
+                }
+                numberOfPropertiesLeftToResolve = unresolvedPropertyNames.size();
+                unresolvedPropertyIterator = unresolvedPropertyNames.iterator();
+            }
+
+        }
+        return resolvedProperties;
+    }
+
+    /**
+     * Gets an unresolved property name from a property variable expression
+     * by stripping the preceding '${' and trailing '}'.
+     *
+     * @param variableExpression the full property variable expression
+     * @return property name
+     */
+    private static String getPropertyNameFromExpression(String variableExpression) {
+        final int propertyStartIndex = variableExpression.lastIndexOf('{') + 1;
```

Comment:
javadoc says that method is `stripping the preceding '${'` while code is only looking for `{`. So it will extract value from text {abc} while it is not technically a property

Reply:
`getPropertyNameFromExpression()` will only ever be passed a string that matches `PROPERTY_VARIABLE_PATTERN` since calls to `getPropertyNameFromExpression()` are guarded by `matcher.find()`.